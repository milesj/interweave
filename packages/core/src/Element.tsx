import React from 'react';

export interface ElementProps {
	[prop: string]: unknown;
	className?: string;
	children?: React.ReactNode;
	selfClose?: boolean;
	tagName: string;
}

export function Element({
	className,
	children = null,
	selfClose = false,
	tagName,
	...props
}: ElementProps) {
	const Tag = tagName as 'span';

	return selfClose ? (
		<Tag className={className} {...props} />
	) : (
		<Tag className={className} {...props}>
			{children}
		</Tag>
	);
}
