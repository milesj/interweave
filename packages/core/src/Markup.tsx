import React, { useMemo } from 'react';
import { Parser, ParserProps } from './Parser';

export interface MarkupProps extends ParserProps {
	/** Content that may contain HTML to safely render. */
	content?: string | null;
	/** Content to render when the `content` prop is empty. */
	emptyContent?: React.ReactNode;
}

export function Markup(props: MarkupProps) {
	const { content, emptyContent } = props;

	const mainContent = useMemo(
		() => new Parser(content ?? '', props).parse(),
		// Do not include `props` as we only want to re-render on content changes
		// eslint-disable-next-line react-hooks/exhaustive-deps
		[content],
	);

	// eslint-disable-next-line react/jsx-no-useless-fragment
	return <>{mainContent ?? emptyContent}</>;
}
