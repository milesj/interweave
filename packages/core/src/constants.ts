/* eslint-disable no-bitwise, no-magic-numbers, sort-keys */

import { ConfigMap, FilterMap, NodeConfig } from './types';

// https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories
export const TYPE_FLOW = 1;
export const TYPE_SECTION = 1 << 1;
export const TYPE_HEADING = 1 << 2;
export const TYPE_PHRASING = 1 << 3;
export const TYPE_EMBEDDED = 1 << 4;
export const TYPE_INTERACTIVE = 1 << 5;
export const TYPE_PALPABLE = 1 << 6;

// https://developer.mozilla.org/en-US/docs/Web/HTML/Element
const tagConfigs: { [tagName: string]: Partial<NodeConfig> } = {
  a: {
    content: TYPE_FLOW | TYPE_PHRASING,
    self: false,
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_INTERACTIVE | TYPE_PALPABLE,
  },
  address: {
    invalid: [
      'h1',
      'h2',
      'h3',
      'h4',
      'h5',
      'h6',
      'address',
      'article',
      'aside',
      'section',
      'div',
      'header',
      'footer',
    ],
    self: false,
  },
  audio: {
    children: ['track', 'source'],
  },
  br: {
    type: TYPE_FLOW | TYPE_PHRASING,
    void: true,
  },
  body: {
    content:
      TYPE_FLOW |
      TYPE_SECTION |
      TYPE_HEADING |
      TYPE_PHRASING |
      TYPE_EMBEDDED |
      TYPE_INTERACTIVE |
      TYPE_PALPABLE,
  },
  button: {
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_INTERACTIVE | TYPE_PALPABLE,
  },
  caption: {
    content: TYPE_FLOW,
    parent: ['table'],
  },
  col: {
    parent: ['colgroup'],
    void: true,
  },
  colgroup: {
    children: ['col'],
    parent: ['table'],
  },
  details: {
    children: ['summary'],
    type: TYPE_FLOW | TYPE_INTERACTIVE | TYPE_PALPABLE,
  },
  dd: {
    content: TYPE_FLOW,
    parent: ['dl'],
  },
  dl: {
    children: ['dt', 'dd'],
    type: TYPE_FLOW,
  },
  dt: {
    content: TYPE_FLOW,
    invalid: ['footer', 'header'],
    parent: ['dl'],
  },
  figcaption: {
    content: TYPE_FLOW,
    parent: ['figure'],
  },
  footer: {
    invalid: ['footer', 'header'],
  },
  header: {
    invalid: ['footer', 'header'],
  },
  hr: {
    type: TYPE_FLOW,
    void: true,
  },
  img: {
    void: true,
  },
  li: {
    content: TYPE_FLOW,
    parent: ['ul', 'ol', 'menu'],
  },
  main: {
    self: false,
  },
  ol: {
    children: ['li'],
    type: TYPE_FLOW,
  },
  picture: {
    children: ['source', 'img'],
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_EMBEDDED,
  },
  rb: {
    parent: ['ruby', 'rtc'],
  },
  rp: {
    parent: ['ruby', 'rtc'],
  },
  rt: {
    content: TYPE_PHRASING,
    parent: ['ruby', 'rtc'],
  },
  rtc: {
    content: TYPE_PHRASING,
    parent: ['ruby'],
  },
  ruby: {
    children: ['rb', 'rp', 'rt', 'rtc'],
  },
  source: {
    parent: ['audio', 'video', 'picture'],
    void: true,
  },
  summary: {
    content: TYPE_PHRASING,
    parent: ['details'],
  },
  table: {
    children: ['caption', 'colgroup', 'thead', 'tbody', 'tfoot', 'tr'],
    type: TYPE_FLOW,
  },
  tbody: {
    parent: ['table'],
    children: ['tr'],
  },
  td: {
    content: TYPE_FLOW,
    parent: ['tr'],
  },
  tfoot: {
    parent: ['table'],
    children: ['tr'],
  },
  th: {
    content: TYPE_FLOW,
    parent: ['tr'],
  },
  thead: {
    parent: ['table'],
    children: ['tr'],
  },
  tr: {
    parent: ['table', 'tbody', 'thead', 'tfoot'],
    children: ['th', 'td'],
  },
  track: {
    parent: ['audio', 'video'],
    void: true,
  },
  ul: {
    children: ['li'],
    type: TYPE_FLOW,
  },
  video: {
    children: ['track', 'source'],
  },
  wbr: {
    type: TYPE_FLOW | TYPE_PHRASING,
    void: true,
  },
};

function createConfigBuilder(config: Partial<NodeConfig>): (tagName: string) => void {
  return (tagName: string) => {
    tagConfigs[tagName] = {
      ...config,
      ...tagConfigs[tagName],
    };
  };
}

['address', 'main', 'div', 'figure', 'p', 'pre'].forEach(
  createConfigBuilder({
    content: TYPE_FLOW,
    type: TYPE_FLOW | TYPE_PALPABLE,
  }),
);

[
  'abbr',
  'b',
  'bdi',
  'bdo',
  'cite',
  'code',
  'data',
  'dfn',
  'em',
  'i',
  'kbd',
  'mark',
  'q',
  'ruby',
  'samp',
  'strong',
  'sub',
  'sup',
  'time',
  'u',
  'var',
].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_PALPABLE,
  }),
);

['p', 'pre'].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PALPABLE,
  }),
);

['s', 'small', 'span', 'del', 'ins'].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_PHRASING,
  }),
);

['article', 'aside', 'footer', 'header', 'nav', 'section', 'blockquote'].forEach(
  createConfigBuilder({
    content: TYPE_FLOW,
    type: TYPE_FLOW | TYPE_SECTION | TYPE_PALPABLE,
  }),
);

['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].forEach(
  createConfigBuilder({
    content: TYPE_PHRASING,
    type: TYPE_FLOW | TYPE_HEADING | TYPE_PALPABLE,
  }),
);

['audio', 'canvas', 'iframe', 'img', 'video'].forEach(
  createConfigBuilder({
    type: TYPE_FLOW | TYPE_PHRASING | TYPE_EMBEDDED | TYPE_PALPABLE,
  }),
);

// Disable this map from being modified
export const TAGS: ConfigMap = Object.freeze(tagConfigs);

// Tags that should never be allowed, even if the allow list is disabled
export const BANNED_TAG_LIST = [
  'applet',
  'base',
  'body',
  'command',
  'embed',
  'frame',
  'frameset',
  'head',
  'html',
  'link',
  'meta',
  'noscript',
  'object',
  'script',
  'style',
  'title',
];

export const ALLOWED_TAG_LIST = Object.keys(TAGS).filter(
  (tag) => tag !== 'canvas' && tag !== 'iframe',
);

// Filters apply to HTML attributes
export const FILTER_ALLOW = 1;
export const FILTER_DENY = 2;
export const FILTER_CAST_NUMBER = 3;
export const FILTER_CAST_BOOL = 4;
export const FILTER_NO_CAST = 5;

// Attributes not listed here will be denied
// https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
export const ATTRIBUTES: FilterMap = Object.freeze({
  alt: FILTER_ALLOW,
  cite: FILTER_ALLOW,
  class: FILTER_ALLOW,
  colspan: FILTER_CAST_NUMBER,
  controls: FILTER_CAST_BOOL,
  datetime: FILTER_ALLOW,
  default: FILTER_CAST_BOOL,
  disabled: FILTER_CAST_BOOL,
  dir: FILTER_ALLOW,
  height: FILTER_ALLOW,
  href: FILTER_ALLOW,
  id: FILTER_ALLOW,
  kind: FILTER_ALLOW,
  label: FILTER_ALLOW,
  lang: FILTER_ALLOW,
  loading: FILTER_ALLOW,
  loop: FILTER_CAST_BOOL,
  media: FILTER_ALLOW,
  muted: FILTER_CAST_BOOL,
  poster: FILTER_ALLOW,
  role: FILTER_ALLOW,
  rowspan: FILTER_CAST_NUMBER,
  scope: FILTER_ALLOW,
  sizes: FILTER_ALLOW,
  span: FILTER_CAST_NUMBER,
  start: FILTER_CAST_NUMBER,
  style: FILTER_NO_CAST,
  src: FILTER_ALLOW,
  srclang: FILTER_ALLOW,
  srcset: FILTER_ALLOW,
  target: FILTER_ALLOW,
  title: FILTER_ALLOW,
  type: FILTER_ALLOW,
  width: FILTER_ALLOW,
});

// Attributes to camel case for React props
export const ATTRIBUTES_TO_PROPS: { [key: string]: string } = Object.freeze({
  class: 'className',
  colspan: 'colSpan',
  datetime: 'dateTime',
  rowspan: 'rowSpan',
  srclang: 'srcLang',
  srcset: 'srcSet',
});
