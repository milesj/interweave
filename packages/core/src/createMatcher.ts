import { CommonInternals, Node, PassthroughProps, TagName } from './types';

// Result from the match process
export interface MatchResult {
	index: number;
	length: number;
	match: string;
	matches: string[];
	valid: boolean;
	value: string;
	void: boolean;
}

// Params returned from `onMatch` that are passed to the factory
export interface MatchParams {
	[key: string]: unknown;
	match?: string;
}

export type OnMatch<Match extends MatchParams, Props extends object> = (
	result: MatchResult,
	props: Props,
) => Match | null;

export interface MatcherOptions<
	Match extends MatchParams,
	Props extends object,
	Config extends object,
> extends CommonInternals<Props, Config> {
	greedy?: boolean;
	tagName: TagName;
	void?: boolean;
	onMatch: OnMatch<Match, Props>;
}

export interface MatcherFactoryData<
	Match extends MatchParams,
	Props extends object,
	Config extends object,
> {
	config: Config;
	params: Match;
	props: Props;
}

export type MatcherFactory<
	Match extends MatchParams,
	Props extends object,
	Config extends object,
> = (
	data: MatcherFactoryData<Match, Props, Config>,
	children: Node | null,
	key: number,
) => React.ReactElement;

export interface Matcher<Match extends MatchParams, Props extends object, Config extends object>
	extends CommonInternals<Props, Config> {
	extend: (
		config?: Partial<Config>,
		factory?: MatcherFactory<Match, Props, Config> | null,
	) => Matcher<Match, Props, Config>;
	factory: MatcherFactory<Match, Props, Config>;
	greedy: boolean;
	match: (value: string, props: Props) => (MatchResult & { params: Match }) | null;
	tagName: TagName;
}

export function createMatcher<
	Match extends MatchParams,
	Props extends object = PassthroughProps,
	Config extends object = {},
>(
	pattern: RegExp | string,
	factory: MatcherFactory<Match, Props, Config>,
	options: MatcherOptions<Match, Props, Config>,
): Matcher<Match, Props, Config> {
	return {
		config: options.config,
		extend(customConfig, customFactory) {
			return createMatcher(pattern, customFactory ?? factory, {
				...options,
				// @ts-expect-error Allow generics merge
				config: {
					...options.config,
					...customConfig,
				},
			});
		},
		factory,
		greedy: options.greedy ?? false,
		match(value, props) {
			const matches = value.match(pattern instanceof RegExp ? pattern : new RegExp(pattern, 'i'));

			if (!matches) {
				return null;
			}

			const result: MatchResult = {
				index: matches.index!,
				length: matches[0].length,
				match: matches[0],
				matches,
				valid: true,
				value,
				void: options.void ?? false,
			};

			const params = options.onMatch(result, props);

			// Allow callback to intercept the result
			if (params === null) {
				return null;
			}

			// Allow callback to replace the matched content
			if ('match' in params && params.match) {
				result.match = params.match;
			}

			return {
				params,
				...result,
			};
		},
		onAfterParse: options.onAfterParse,
		onBeforeParse: options.onBeforeParse,
		tagName: options.tagName,
	};
}
