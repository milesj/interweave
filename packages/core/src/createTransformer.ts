import { CommonInternals, Node, PassthroughProps, TagName, WildTagName } from './types';

export type InferElement<K> = K extends '*'
	? HTMLElement
	: K extends keyof HTMLElementTagNameMap
	? HTMLElementTagNameMap[K]
	: HTMLElement;

export type TransformerFactory<Element, Props extends object> = (
	element: Element,
	props: Props,
	children: Node[],
) => Element | React.ReactElement | null | undefined | void;

export interface TransformerOptions<Props extends object, Config extends object>
	extends CommonInternals<Props, Config> {
	tagName?: TagName;
}

export interface Transformer<Element, Props extends object, Config extends object>
	extends CommonInternals<Props, Config> {
	extend: (
		config?: Partial<Config>,
		factory?: TransformerFactory<Element, Props>,
	) => Transformer<Element, Props, Config>;
	factory: TransformerFactory<Element, Props>;
	tagName: WildTagName;
}

export function createTransformer<
	K extends WildTagName,
	Props extends object = PassthroughProps,
	Config extends object = {},
>(
	tagName: K,
	factory: TransformerFactory<InferElement<K>, Props>,
	options: TransformerOptions<Props, Config> = {},
): Transformer<InferElement<K>, Props, Config> {
	return {
		config: options.config,
		extend(customConfig, customFactory) {
			return createTransformer<K, Props, Config>(tagName, customFactory ?? factory, {
				...options,
				config: {
					...options.config!,
					...customConfig,
				},
			});
		},
		factory,
		onAfterParse: options.onAfterParse,
		onBeforeParse: options.onBeforeParse,
		tagName: options.tagName ?? tagName,
	};
}
