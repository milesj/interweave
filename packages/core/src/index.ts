/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './constants';
export * from './createMatcher';
export * from './createTransformer';
export * from './Element';
export * from './Interweave';
export * from './Markup';
export * from './Parser';
export * from './types';
