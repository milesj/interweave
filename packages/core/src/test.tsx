import React from 'react';
import { createMatcher, createTransformer, Element, TagConfig, TAGS } from '.';

export const TOKEN_LOCATIONS = [
	'no tokens',
	'{token}',
	' {token} ',
	'{token} pattern at beginning',
	'pattern at end {token}',
	'pattern in {token} middle',
	'{token} pattern at beginning and end {token}',
	'{token} pattern on {token} all sides {token}',
	'pattern {token} used {token} multiple {token} times',
	'tokens next {token} {token} {token} to each other',
	'token next to {token}, a comma',
	'token by a period {token}.',
	'token after a colon: {token}',
	'token after a\n{token} new line',
	'token before a {token}\n new line',
	'token surrounded by ({token}) parenthesis',
	// 'tokens without {token}{token}{token} spaces',
];

export const SOURCE_PROP = {
	compact: false,
	locale: 'en',
	version: '0.0.0',
} as const;

export const VALID_EMOJIS = [
	['1F621', '😡', ':rage:', '>:/'],
	['1F468-200D-1F469-200D-1F467-200D-1F466', '👨‍👩‍👧‍👦', ':family_mwgb:'],
	['1F1FA-1F1F8', '🇺🇸', ':flag_us:'],
	['1F63A', '😺', ':smiley_cat:'],
	['1F3EF', '🏯', ':japanese_castle:'],
	['1F681', '🚁', ':helicopter:'],
	['1F469-200D-2764-FE0F-200D-1F468', '👩‍❤️‍👨', ':couple_with_heart_mw:'],
	['1F1E7-1F1F4', '🇧🇴', ':flag_bo:'],
	['1F468-200D-1F468-200D-1F466', '👨‍👨‍👦', ':family_mmb:'],
	['1F3C0', '🏀', ':basketball:'],
];

export function createExpectedToken<T>(
	value: T,
	factory: (value: T, count: number) => React.ReactNode,
	index: number,
	join: boolean = false,
): React.ReactNode {
	if (index === 0) {
		return TOKEN_LOCATIONS[0];
	}

	let count = -1;
	const tokens = TOKEN_LOCATIONS[index]
		.split(/({token})/)
		.map((row) => {
			if (row === '{token}') {
				count += 1;

				return factory(value, count);
			}

			return row;
		})
		.filter(Boolean);

	return join ? tokens.join('') : tokens;
}

export const MOCK_MARKUP = `<main role="main">
  Main content
  <div>
    <a href="#">Link</a>
    <span class="foo">String</span>
  </div>
</main>
<aside id="sidebar">
  Sidebar content
</aside>`;

export const MOCK_INVALID_MARKUP = `<div bgcolor="black">
  <font color="red">Outdated font.</font>
  <script type="text/javascript"></script>
  <p align="center">More text <strike>with outdated stuff</strike>.</p>
</div>`;

export const parentConfig: TagConfig = {
	children: [],
	content: 0,
	invalid: [],
	parent: [],
	self: true,
	tagName: 'div',
	type: 0,
	void: false,
	...TAGS.div,
};

export const codeFooMatcher = createMatcher(
	/\[foo]/,
	({ params }, children, key) => (
		<Element key={key} customprop={params.customprop} tagName="span">
			{String(children).toUpperCase()}
		</Element>
	),
	{
		onMatch: () => ({
			match: 'foo',
			customprop: 'foo',
		}),
		tagName: 'span',
	},
);

export const codeBarMatcher = createMatcher(
	/\[bar]/,
	({ params }, children, key) => (
		<Element key={key} customprop={params.customprop} tagName="span">
			{String(children).toUpperCase()}
		</Element>
	),
	{
		onMatch: () => ({
			match: 'bar',
			customprop: 'bar',
		}),
		tagName: 'span',
	},
);

export const codeBazMatcher = createMatcher(
	/\[baz]/,
	({ params }, children, key) => (
		<Element key={key} customprop={params.customprop} tagName="span">
			{String(children).toUpperCase()}
		</Element>
	),
	{
		onMatch: () => ({
			match: 'baz',
			customprop: 'baz',
		}),
		tagName: 'span',
	},
);

export const mdBoldMatcher = createMatcher(
	/\*\*([^*]+)\*\*/u,
	(params, children, key) => <b key={key}>{children}</b>,
	{
		onMatch: ({ matches }) => ({
			match: matches[1],
		}),
		tagName: 'b',
	},
);

export const mdItalicMatcher = createMatcher(
	/_([^_]+)_/u,
	(params, children, key) => <i key={key}>{children}</i>,
	{
		onMatch: ({ matches }) => ({
			match: matches[1],
		}),
		tagName: 'i',
	},
);

export const mockMatcher = createMatcher(
	/div/,
	({ props }, children, key) => (
		<div {...props} key={key}>
			{children}
		</div>
	),
	{
		onMatch: () => null,
		tagName: 'div',
	},
);

export const linkTransformer = createTransformer('a', (element) => {
	element.setAttribute('target', '_blank');

	const href = element.getAttribute('href');

	if (href) {
		element.setAttribute('href', href.replace('foo.com', 'bar.net') || '');
	}
});

export const mockTransformer = createTransformer('*', () => {});
