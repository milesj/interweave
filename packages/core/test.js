module.exports = require('./lib/test.js');
