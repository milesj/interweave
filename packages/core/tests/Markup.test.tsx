import React from 'react';
import { render } from 'rut-dom';
import { Markup, MarkupProps } from '../src/Markup';
import { MOCK_MARKUP } from '../src/test';

const options = { log: false, reactElements: false };

describe('Markup', () => {
	it('allows empty `content` to be passed', () => {
		const { root } = render<MarkupProps>(<Markup content={null} />);

		expect(root.debug(options)).toMatchSnapshot();
	});

	it('will render the `emptyContent` if no content exists', () => {
		const empty = <div>Foo</div>;
		const { root } = render<MarkupProps>(<Markup content="" emptyContent={empty} />);

		expect(root.debug(options)).toMatchSnapshot();
	});

	it('parses the entire document starting from the body', () => {
		const { root } = render<MarkupProps>(<Markup content={MOCK_MARKUP} />);

		expect(root.debug(options)).toMatchSnapshot();
	});

	it('converts line breaks', () => {
		const { root } = render<MarkupProps>(<Markup content={'Foo\nBar'} />);

		expect(root.debug(options)).toMatchSnapshot();
	});

	it('doesnt convert line breaks', () => {
		const { root } = render<MarkupProps>(<Markup disableLineBreaks content={'Foo\nBar'} />);

		expect(root.debug(options)).toMatchSnapshot();
	});

	it('doesnt convert line breaks if it contains HTML', () => {
		const { root } = render<MarkupProps>(<Markup content={'Foo\n<br/>Bar'} />);

		expect(root.debug(options)).toMatchSnapshot();
	});
});
