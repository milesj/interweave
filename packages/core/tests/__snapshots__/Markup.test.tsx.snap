// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markup allows empty \`content\` to be passed 1`] = `""`;

exports[`Markup converts line breaks 1`] = `
"Foo
<br />
Bar"
`;

exports[`Markup doesnt convert line breaks 1`] = `
"Foo
Bar"
`;

exports[`Markup doesnt convert line breaks if it contains HTML 1`] = `
"Foo

<br />
Bar"
`;

exports[`Markup parses the entire document starting from the body 1`] = `
"<main role=\\"main\\">
  
  Main content
  
  <div>
    
    
    <a href=\\"#\\">
      Link
    </a>
    
    
    <span className=\\"foo\\">
      String
    </span>
    
  
  </div>
  

</main>


<aside id=\\"sidebar\\">
  
  Sidebar content

</aside>"
`;

exports[`Markup will render the \`emptyContent\` if no content exists 1`] = `"<div>Foo</div>"`;
